sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala

```diff
@@ -56,117 +52,247 @@ case class SortMergeJoin(
+        matchJoinKey = null
+        buildMatches = null
+      } else if (matchJoinKey != null && keyOrdering.compare(streamedRowKey, matchJoinKey) == 0) {
+        // Matches the current group, so do nothing.
+      } else {
+        // The streamed row does not match the current group.
+        matchJoinKey = null
+        buildMatches = null
+        if (buildRow != null) {
+          // The build iterator could still contain matching rows, so we'll need to walk through it
+          // until we either find matches or pass where they would be found.
+          var comp =
+            if (buildRowKey.anyNull) 1 else keyOrdering.compare(streamedRowKey, buildRowKey)
+          while (comp > 0 && advanceBuild()) {
+            comp = if (buildRowKey.anyNull) 1 else keyOrdering.compare(streamedRowKey, buildRowKey)
+          }
+          if (comp == 0) {
+            // We have found matches, so buffer them (this updates matchJoinKey)
+            bufferMatchingBuildRows()
+          } else {
+            // We have overshot the position where the row would be found, hence no matches.
           }
-          rightMatches != null && rightMatches.size > 0
         }
       }
+      // If there is a streamed input, then we always return true
+      true
+    } else {
+      // End of streamed input, hence no more results.
+      false
     }
   }
+
+  def getStreamedRow: InternalRow = streamedRow
+  def getBuildMatches: CompactBuffer[InternalRow] = buildMatches
+
+  // --- Private methods --------------------------------------------------------------------------
+
+  /**
+   * Advance the streamed iterator and compute the new row's join key.
+   * @return true if the streamed iterator returned a row and false otherwise.
+   */
+  private def advancedStreamed(): Boolean = {
+    if (streamedIter.hasNext) {
+      streamedRow = streamedIter.next()
+      streamedRowKey = streamedKeyGenerator(streamedRow)
+      true
+    } else {
+      streamedRow = null
+      streamedRowKey = null
+      false
+    }
+  }
+
+  /**
+   * Advance the build iterator and compute the new row's join key.
+   * @return true if the build iterator returned a row and false otherwise.
+   */
+  private def advanceBuild(): Boolean = {
+    if (buildIter.hasNext) {
+      buildRow = buildIter.next()
+      buildRowKey = buildKeyGenerator(buildRow)
+      true
+    } else {
+      buildRow = null
+      buildRowKey = null
+      false
+    }
+  }
+
+  /**
+   * Called when the streamed and build join keys match in order to buffer the matching build rows.
+   */
+  private def bufferMatchingBuildRows(): Unit = {
+    assert(streamedRowKey != null)
+    assert(!streamedRowKey.anyNull)
+    assert(buildRowKey != null)
+    assert(!buildRowKey.anyNull)
+    assert(keyOrdering.compare(streamedRowKey, buildRowKey) == 0)
+    matchJoinKey = streamedRowKey.copy()
+    buildMatches = new CompactBuffer[InternalRow]
```

Comment:
Good question; I think that there's a time / space trade-off at place here depending on how many of these buffers you expect to create. I haven't benchmarked this yet, but I'm fairly certain that re-creating the `CompactBuffer` here is going to have an allocation hit compared to just using an ArrayBuffer, especially given the costs of the extra branches when adding elements, etc.